### PR TITLE
pypy parity: index/slice coercion, sequence concat, in-place operators, overflow guards

### PR DESCRIPTION
**Pull request by youknowone**
Drains the remaining low-risk pypy-parity items from the 3-way (cpython/pypy/pyre) sweep — the integer-coercion, sequence-concatenation, in-place-operator, and numeric-overflow paths. Each behavior was probed against the installed `pypy3` (7.3.20, the `check.py` oracle) and now matches byte-for-byte. Where cpython and pypy disagree (so a `check.py` baseline test cannot cover it), the behavior was verified by hand against `pypy3`.

### `objspace: remap __index__ TypeError; ord names bytes noun`
Carries over the two PR #552 review fixes that had not yet landed:
- `getindex_w` remaps a `TypeError` raised by a subscript's `__index__` (a non-int return, or a `TypeError` from inside `__index__`) to `<descr> indices must be integers or slices, not '<type>'`; a `ValueError` still propagates. Applied to the list/tuple/str/bytes get, list/bytearray set, and delitem sites.
- `ord(b"ab")` reports `bytes of length 2` while `bytearray(b"ab")` keeps `string of length 2`.

### `pypy parity: index coercion and sequence-concat errors`
- **`builtins::getindex_w`** — an out-of-word overflow now clamps to `i64::MIN` for a negative value (was always `i64::MAX`), so a large-negative slice bound floors to `0` instead of collapsing to the tail: `[1,2,3][-(10**30):]` → `[1, 2, 3]`.
- **`normalize_slice` / `sliceobject::adapt_lower_bound`** — fold a negative index by the length with `saturating_add` so an `i64::MIN` bound does not overflow before flooring at 0.
- **`str_slice_args` (`startswith`/`endswith`)** — take the `start`/`end` bounds through `adapt_lower_bound(eval_slice_index(...))`, so a non-index bound raises `slice indices must be integers or None or have an __index__ method`, an `__index__` object is honoured, and a positive bound is not upper-clamped.
- **`builtin_round`** — take `ndigits` through `getindex_w` (float: clamping; int: `space.index`); a huge-magnitude negative `ndigits` short-circuits to `0` rather than building an astronomical power of ten (`round(3.14159, 2**63)` → `3.14159`).
- **`index_to_bigint` (`hex`/`oct`/`bin`)** — coerce through `space.index` so a non-int `__index__` return raises `__index__ returned non-int (type X)`.
- **formatting `number_arg_decimal` / `number_arg_integer` (`%d`/`%i`/`%o`/`%x`)** — remap a `TypeError` from the numeric decoder to the operand-type error naming the original argument (`%d format: a real number is required, not BadIdx`).
- **`list`/`tuple`/`bytes`/`bytearray` `__add__` and `descroperation::add`** — return `NotImplemented` for a non-sequence operand so the operator raises the generic `unsupported operand type(s) for +`, dropping the cpython-only `can only concatenate` message; `bytearray.__iadd__` on a non-buffer now raises `a bytes-like object is required, not 'X'`.

### `fix warnings`
Removes three unused-binding warnings in the rtyper (`classdesc.rs`, `rclass.rs`, `rpbc.rs`).

### `operator/set: in-place operator functions and set slots`
- The `operator` module gains the in-place functions `iadd`, `isub`, `imul`, `imatmul`, `ifloordiv`, `imod`, `itruediv`, `ipow`, `ilshift`, `irshift`, `iand`, `ior`, `ixor` (each dispatching the matching in-place binary op) and `iconcat` (which requires both operands to be subscriptable, else raises `'X' object can't be concatenated`). The `__iadd__` → `iadd` alias table now resolves to real callables.
- `set` gains the `__isub__`, `__iand__`, `__ior__`, `__ixor__` slots, so `s &= t` and `operator.iand(s, t)` update in place; a non-set right operand yields `NotImplemented`.

### `int: raise MemoryError from an unallocatably-large left shift`
- `1 << (10**18)` drove the big-int `<<` into the infallible global allocator and aborted the process. A new `checked_bigint_lshift` pre-flights a fallible reservation of the result's 64-bit limb count and raises a catchable `MemoryError` instead; `int` and `long` left shifts route through it. The shift count is carried as `u64` so a 32-bit `usize` target (wasm) does not truncate a `10**18` shift.

### `float/complex pow: over-range int operand overflows`
- A `float`/`complex` power coerces each `int` operand to a double before the power is computed, so an over-range `int` base or exponent raises `OverflowError` up front — even `1.0 ** huge`, which never reaches the arithmetic, matching `float(huge)`. In `float.__pow__`/`__rpow__` the check runs before the ternary-modulus rejection, so `pow(2.0, huge, 5)` raises `OverflowError` from the coercion rather than `TypeError` from the modulus.

## Verification
- Probes (custom `__index__` objects, huge/negative bounds, non-seq concat, `%`-format, in-place set/operator ops, `1 << 10**18`, over-range float/complex pow) are byte-identical to `pypy3` across all cases; the cpython≠pypy edges (3-arg pow ordering, lshift OOM message) were hand-verified against `pypy3`.
- Normal slice / concat / `hex` / `round` / `%`-format / `startswith` / in-place-op / small-shift / ordinary-pow paths are byte-identical to `pypy3` (no regression).
- New synth tests: `operator_set_inplace_ops`, `int_lshift_memoryerror`, `float_pow_overflow_exp` (each restricted to cpython==pypy cases).
- `check.py`: dynasm 190/190, cranelift 190/190, wasm 189/189 (3/3 backends).

## Deferred (own follow-up slices)
- **`int.__lshift__('a', 1)`** returns `NotImplemented` where pypy raises `TypeError "'int' object expected, got 'str' instead"` — needs a systematic unbound-descriptor self-type check across every typed-self builtin; cpython≠pypy so not baseline-testable.
- **`x &= y` augmented-assignment error symbol** — a type mismatch reports the binary-op symbol `&` rather than `&=` (e.g. `int += str`). Needs an `Option`-returning binop path so the augmented-assign layer can re-emit with the `X=` symbol.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for `operator.iconcat`.
  * Expanded set and dictionary operations, including in-place set operators and dictionary-view set operations.

* **Bug Fixes**
  * Improved slice-bound handling, index coercion, and boundary behavior for string and sequence operations.
  * Corrected error messages and exception propagation for hashing, unhashable values, overflow, memory limits, and invalid operands.
  * Improved behavior for `round()`, `pow()`, `ord()`, and set/dictionary updates.

* **Tests**
  * Added benchmarks covering dictionary, set, operator, numeric, and string edge cases.

### CODE REVIEW

File: pyre/pyre-interpreter/src/type_methods.rs [hunk 1]
```diff
@@ -888,44 +888,41 @@ pub fn str_method_endswith(args: &[PyObjectRef]) -> Result<PyObjectRef, crate::P
...
         0
     };
     let end = if args.len() >= 4 && !unsafe { pyre_object::is_none(args[3]) } {
-        let v = unsafe { pyre_object::w_int_get_value(args[3]) };
-        if v < 0 {
-            (char_len + v).max(0) as usize
-        } else {
-            (v as usize).min(char_len as usize)
-        }
+        crate::sliceobject::adapt_lower_bound(char_len, args[3])?
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/type_methods.rs:924]: Confirmed and fixed in \`caf61cb243\`.

\`'abc'.startswith('', 5, 10)\` and \`endswith\` returned \`True\` against both oracles' \`False\`. Rather than clamping \`end\`, this ports the upstream mechanism directly — \`unicodeobject.py:1319 _unwrap_and_compute_idx_params\` turns a \`start\` past the end into \`end_index + 1\`, inverting the window, and only lowers \`end_index\` when \`end < length\`. The byte-space \`start_index > end_index\` test then subsumes the old \`start > end\` check, so \`''.endswith('', 1, 0)\` stays \`False\` too.

Guarded by the new synth test \`str_startswith_bounds.py\` (27 lines, cpython == pypy == pyre).

— *commented by Claude*

File: pyre/pyre-interpreter/src/objspace/descroperation.rs [hunk 2]
```diff
@@ -82,6 +82,36 @@ fn bigint_lshift(a: BigInt, shift: usize) -> BigInt {
...
 
+/// `longobject.py` left shift — the underlying big-int `<<` aborts the process
+/// when the shifted result cannot be allocated, so pre-flight a fallible
+/// reservation of the result's 64-bit limbs and raise a catchable MemoryError
+/// instead. The probe uses the same global allocator and a limb count that
+/// upper-bounds the shift's own allocation, so it fails exactly when the real
+/// shift would.
+fn checked_bigint_lshift(a: BigInt, shift: u64) -> Result<BigInt, PyError> {
+    let result_bits = a.bits().saturating_add(shift);
+    let limbs = (result_bits / 64).saturating_add(2);
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/objspace/descroperation.rs:98]: Confirmed and fixed in \`caf61cb243\`.

\`0 << 10**18\` raised MemoryError; both oracles return \`0\`. \`rbigint.py:1323 lshift\` returns self when \`int_other == 0 or selfsign == 0\`, so neither a zero operand nor a zero shift allocates however large the count — that guard is now the first thing \`checked_bigint_lshift\` does, which covers both the int and long callers. \`1 << 10**18\` still raises MemoryError.

Guarded by \`str_startswith_bounds.py\`.

— *commented by Claude*

File: pyre/pyre-interpreter/src/typedef.rs [hunk 3]
```diff
@@ -8173,6 +8103,9 @@ fn float_dunder_pow(args: &[PyObjectRef]) -> Result<PyObjectRef, crate::PyError>
     crate::type_methods::arity_pow(args)?;
     let b = args[1];
     if unsafe { pyre_object::pyobject::is_float(b) || pyre_object::pyobject::is_int_or_long(b) } {
+        // The operand is coerced to a double first, so an over-range int
+        // raises OverflowError before the ternary modulus is rejected.
+        unsafe { crate::objspace::descroperation::reject_pow_operand_overflow(b)? };
         float_pow_reject_modulus(args)?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/typedef.rs:9906]: Not applying this one — the requested behaviour is CPython's, not PyPy's, and pyre targets PyPy.

\`\`\`
pow(1.0, 10**1000, 3)
  pypy3 7.3.20 : OverflowError: int too large to convert to float
  cpython 3.14 : TypeError: pow() 3rd argument not allowed unless all arguments are integers
  pyre         : OverflowError: int too large to convert to float
\`\`\`

PyPy converts the exponent before it rejects the modulus, so the OverflowError wins; the current ordering is deliberate and matches it. \`float.__pow__(1.0, 10**1000, 3)\` and \`pow(2.0, 10**1000, 5)\` behave the same way in both pypy3 and pyre. \`pow(2.0, 3, 5)\` still raises the ternary TypeError.

This is also why these lines cannot be added to a synth test: check.py's baseline gate requires cpython == pypy, and here they disagree — so they are hand-verified against pypy3 instead.

— *commented by Claude*

File: pyre/pyre-interpreter/src/typedef.rs [hunk 4]
```diff
@@ -4389,111 +4388,43 @@ fn dict_view_isdisjoint(
...
-                }
-            }
-            out
-        }
-    };
-    Ok(pyre_object::w_set_from_items(&result_items))
+    let w_set_type = crate::typedef::gettypeobject(&pyre_object::setobject::SET_TYPE);
+    let w_set = crate::call::call_function_impl_result(w_set_type, &[lhs])?;
+    let method = crate::baseobjspace::getattr_str(w_set, methname)?;
+    crate::call::call_function_impl_result(method, &[rhs])?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/typedef.rs:5728]: Half right, and the real half is fixed in \`caf61cb243\`.

The one-shot example already worked — \`d={1:0,2:0}; d.keys() & iter([1,2])\` gives \`[1, 2]\` in pyre, matching both oracles, because the view op materialises its operand before the membership pass. The duplicate case was genuinely broken: \`d={1:0}; d.keys() ^ [2,2]\` gave \`[1]\` against \`[1, 2]\`.

Root cause is the same as the \`symmetric_difference_update\` thread below, and fixing that fixed this. Both are guarded by the new \`set_update_materialize_rhs.py\` (19 lines, cpython == pypy == pyre), which covers the one-shot and duplicate cases for \`&\`, \`|\`, \`-\` and \`^\`.

— *commented by Claude*

File: pyre/pyre-interpreter/src/typedef.rs [hunk 5]
```diff
@@ -13867,6 +13937,128 @@ fn set_method_ge(
...
+) -> Result<pyre_object::PyObjectRef, crate::PyError> {
+    if args.is_empty() {
+        return Ok(pyre_object::w_none());
+    }
+    for other in &args[1..] {
+        let other_items = crate::builtins::collect_iterable(*other)?;
+        for item in other_items {
+            // Each element is hashed as it is looked up, so an unhashable one
+            // raises even when self is empty and nothing can match.
+            crate::type_methods::set_discard_checked(args[0], item)?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/typedef.rs]: Confirmed and fixed in \`caf61cb243\`. \`s = {1, 2}; s.difference_update([1, []])\` left \`s\` as \`[2]\`; pypy leaves it \`[1, 2]\`. Now ports \`setobject.py:389 descr_difference_update\` — the operand becomes a set first, so it is hashed and deduped before anything is removed.

Note this one is cpython != pypy (cpython leaves \`[2]\`, pypy \`[1, 2]\`), so it is hand-verified against pypy3 rather than synth-gated.

— *commented by Claude*

File: pyre/pyre-interpreter/src/typedef.rs [hunk 6]
```diff
@@ -13867,6 +13937,128 @@ fn set_method_ge(
...
+            if crate::baseobjspace::eq_w(item, existing)? {
+                present = true;
+                break;
+            }
+        }
+        if present {
+            unsafe { pyre_object::w_set_discard(args[0], item) };
+        } else {
+            crate::type_methods::set_add_checked(args[0], item)?;
+        }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/typedef.rs]: Confirmed and fixed in \`caf61cb243\`. \`s = {1, 2}; s.symmetric_difference_update([3, []])\` left \`s\` as \`[1, 2, 3]\`; pypy leaves it \`[1, 2]\`. Now ports \`setobject.py:497 descr_symmetric_difference_update\` — building the set first both dedupes the operand (fixing the double-toggle in the dict-view thread) and leaves self untouched when it raises.

Worth recording the boundary: \`descr_update:506\` deliberately does NOT materialise, so \`{1, 2}.update([3, []])\` legitimately leaves \`[1, 2, 3]\` in pypy and cpython alike — pyre already matched that and it is kept, with a synth line pinning it.

— *commented by Claude*

File: pyre/pyre-interpreter/src/typedef.rs [hunk 7]
```diff
@@ -13738,18 +13744,82 @@ fn set_method_union(
...
+/// `setobject.py:407 W_SetObject._newobj` / `setobject.py:609
+/// W_FrozensetObject._newobj` — both take ownership of the iterable's
+/// elements; only the resulting class differs, and the intersection reads
+/// back the elements rather than the object, so one set-shaped result
+/// serves both.
+fn set_newobj_items(
+    w_iterable: pyre_object::PyObjectRef,
+) -> Result<Vec<pyre_object::PyObjectRef>, crate::PyError> {
+    let items = crate::builtins::collect_iterable(w_iterable)?;
+    let w_set = crate::type_methods::set_from_items_checked(&items)?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/typedef.rs]: Fixed in `66dac3d8f5` — the finding is correct, and it reached further than the citation.

`set_from_items_checked` → `w_set_from_items_checked` did `let s = w_set_new()` and then held `s` as a raw pointer across every element's `object_key_for_checked`, i.e. across every user `__hash__`, while reading the elements from an unrooted slice. Both halves of what you describe.

Rather than add rooting to that builder, `set_newobj_items` now builds through `builtin_set_from_items`, which already carries exactly the convention you point at in `set_method_update` — root `self` and every element, hash, reload both, store — and returns the reloaded set. That left `set_from_items_checked`, `w_set_from_items_checked` and `w_set_add_checked` with no callers, so they are deleted rather than left in the tree as an unrooted builder for the next caller to reach for.

Verified with the scenario you named: `{a}.intersection([a])` / `{a} & {a}` / `{a}.union([A(), A()])` over 2000 iterations with an `A.__hash__` that allocates to provoke a minor collection — byte-identical to pypy3. `python ./pyre/check.py` is green on dynasm 204/204 and cranelift 204/204.

Two notes for the record:

- The same commit's parent fixes a related defect this rooting work surfaced: the gate hashed each element **twice** (`try_hash_value` for the raise check, then `w_set_add` → `object_key_for` → the `hash_w` hook → the same `try_hash_value`) and **swallowed** an `eq_w` raise from the bucket probe, so `set([EqBoom(), EqBoom()])` returned a 2-element set where `newset`'s `r_dict(space.eq_w, space.hash_w)` (setobject.py:1611) raises. Both are now gated by `pyre/bench/synth/set_key_protocol.py`.
- Still open, tracked separately: the `eq_w` probe is itself a collection point (a user `__eq__` under a hash collision), so `s.len += 1` / `dict_write_barrier(obj)` after `entries.insert` can still write through a relocated header. That hazard is uniform and pre-existing — `w_dict_store_object_strategy_checked` on the `dict.__setitem__` path ships it on main today — so this change does not regress it, but it is not closed either.

— *commented by Claude*

File: pyre/pyre-interpreter/src/typedef.rs [hunk 8]
```diff
@@ -15140,18 +15155,86 @@ fn set_method_union(
...
+/// serves both.
+fn set_newobj_items(
+    w_iterable: pyre_object::PyObjectRef,
+) -> Result<Vec<pyre_object::PyObjectRef>, crate::PyError> {
+    let items = crate::builtins::collect_iterable(w_iterable)?;
+    // Each element's `__hash__` is a collection point that can move both the
+    // temporary set and the elements not yet added, so the build roots them
+    // and reloads across every hash; `builtin_set_from_items` returns the
+    // reloaded set, and reading its elements back allocates nothing.
+    let w_set = crate::builtins::builtin_set_from_items(&items)?;
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/typedef.rs]: Confirmed and fixed — and the finding generalised further than the citation.

You are right that `set_strategy_and_setdata` gives a set operand its own branch (`setobject.py:1619-1621`), and that pyre rebuilt it instead. Measuring it (a `__hash__` that counts calls, one element per set) showed **every** set-to-set line diverging, not just the four named:

| case | pypy | pyre before |
|---|---|---|
| `s.intersection(s)` | 0 | 5 |
| `s.intersection(t)` | 0 | 4 |
| `s.union(t)` | 0 | 2 |
| `s.difference(t)` | 0 | 3 |
| `set(s)` / `frozenset(s)` | 0 | 1 |
| `s.intersection_update(s)` | 1 | 5 |

So the root was not the four operand sites but the primitive under them: `w_set_items` returns `Vec<PyObjectRef>` built from `.keys().map(|key| key.obj)`, i.e. it **drops `key.hash`**, so every rebuild had to re-hash.

Rather than branch on `is_set_or_frozenset` at each call site over the same lossy primitive, the storage layer now mirrors upstream:

- `w_set_copy_storage_from` ← `:1620-1621` + `:963-966 get_storage_copy`, which is `d.copy()` — a **bulk table copy**, not an element-by-element refill, so it calls neither `__hash__` nor `__eq__`.
- `w_set_key_at` / `w_set_contains_key_checked` / `w_set_insert_key_checked` ← `iterkeys_with_hash` / `contains_with_hash` / `setitem_with_hash` (`:1147-1154 _intersect_unwrapped`).
- `w_set_update_from_set` ← `:1400-1403 ObjectSetStrategy.update` (`d_obj.update(d_other)`).

The set methods then take upstream's shape: `descr_union:363` = `copy_real()` + merge, `descr_difference:281` = `copy_real()` + `difference_update`, `descr_intersection:309` seeds with `_newobj` and intersects set operands as they stand.

All set-to-set counts now equal pypy's exactly. Gated in `pyre/bench/synth/set_operand_storage.py` (baseline-clean: cpython and pypy agree on every line).

One genuine divergence remains and is deliberately **not** in this change: `frozenset.__hash__` re-walks on every call because pyre's `W_SetObject` has no `hash` memo field (`descr_hash:624` early-returns, `:643` stores). That is a different mechanism and it changes the GC struct layout, so it is filed separately.

— *commented by Claude*

File: pyre/pyre-interpreter/src/typedef.rs [hunk 9]
```diff
@@ -14803,6 +14748,76 @@ fn init_bytearray_type(ns: PyObjectRef) {
...
+/// frozenset rather than copying; the elements hashed when they entered the
+/// set, so sharing them cannot raise.
+fn convert_set_to_frozenset(w_obj: PyObjectRef) -> Option<PyObjectRef> {
+    unsafe {
+        if !pyre_object::is_set(w_obj) {
+            return None;
+        }
+        Some(pyre_object::w_frozenset_from_items(
+            &pyre_object::w_set_items(w_obj),
+        ))
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/typedef.rs]: Confirmed and fixed. The comment did promise storage sharing while the code did `w_set_items` + `w_frozenset_from_items`, i.e. it rebuilt from bare elements and re-hashed each one — the comment was describing the intent, not the code.

`convert_set_to_frozenset` now allocates the frozenset and calls `w_set_copy_storage_from`, which mirrors `setobject.py:1620-1621 set_strategy_and_setdata` and `:963-966 ObjectSetStrategy.get_storage_copy` (`d.copy()` — a bulk table copy that runs neither `__hash__` nor `__eq__`).

Worth noting the ordering this forces: allocate first, then copy. A `Vec<ObjectKey>` lifted out of the source and held across the allocation would be stale, because `set_object_custom_trace` (pyre-jit/src/eval.rs:564) rewrites the `obj` slots **inside** the table and cannot see a detached copy. Allocating first is also exactly upstream's order (`allocate_instance` then `__init__` → `set_strategy_and_setdata`).

Measured: `frozenset(s)` went 1 → 0 hashes, matching pypy. Gated in `pyre/bench/synth/set_operand_storage.py` (`ctor_frozenset_of_set`).

— *commented by Claude*

File: pyre/pyre-interpreter/src/typedef.rs [hunk 10]
```diff
@@ -15140,18 +15155,86 @@ fn set_method_union(
...
+) -> Vec<pyre_object::PyObjectRef> {
+    let (keep, probe) = if w_set.len() > w_other.len() {
+        (w_other, w_set)
+    } else {
+        (w_set, w_other)
+    };
+    let probe = pyre_object::w_set_from_items(&probe);
+    keep.into_iter()
+        .filter(|&item| unsafe { pyre_object::w_set_contains(probe, item) })
+        .collect()
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/typedef.rs]: Confirmed and fixed. `set_newobj_items` is gone; `set_newobj` now returns the set (as `_newobj:407` does) instead of flattening it to elements, and the intersection carries `ObjectKey`s throughout.

`set_intersect_update` now mirrors `setobject.py:1147-1154 _intersect_unwrapped` directly — walk one side as `(key, keyhash)` pairs, probe the other with `contains_with_hash`, place survivors with `setitem_with_hash`:

```rust
while let Some(key) = w_set_key_at(keep, i) {
    if w_set_contains_key_checked(probe, key)? {
        let key = w_set_key_at(keep, i).expect(...);  // re-read: the probe's eq_w can move it
        w_set_insert_key_checked(result, key)?;
    }
    i += 1;
}
```

Two things fell out of doing it this way. The `w_set_contains` probes became `_checked`, so an `__eq__` raising mid-probe now propagates instead of reading as "not a member" — `{EqBoom()} & {EqBoom()}` raises where it used to return a set. And the walk reads one index at a time rather than collecting: the probe's `eq_w` is a collection point, and the collector rewrites the `obj` slots inside the table in place, so a lifted `Vec` would be left holding stale pointers.

`s.intersection(s)` went 5 → 0 hashes and `s.intersection(t)` 4 → 0, both matching pypy.

— *commented by Claude*

File: pyre/pyre-interpreter/src/typedef.rs [hunk 11]
```diff
@@ -15269,6 +15352,172 @@ fn set_method_ge(
...
+    args: &[pyre_object::PyObjectRef],
+) -> Result<pyre_object::PyObjectRef, crate::PyError> {
+    if args.is_empty() {
+        return Ok(pyre_object::w_none());
+    }
+    for other in &args[1..] {
+        let other_items = set_newobj_items(*other)?;
+        for item in other_items {
+            unsafe { pyre_object::w_set_discard(args[0], item) };
+        }
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/typedef.rs]: Confirmed, and the symmetric-difference half was the more serious of the two — thank you.

`difference_update` did use an unchecked discard, so an `__eq__` raising from the bucket probe was swallowed. It now walks the operand set with `w_set_key_at` and removes through `w_set_discard_key_checked`, under the digest each key already carries (`setobject.py:1058 _difference_update_unwrapped`).

The symmetric-difference path was worse than "does not consider hashes": it was a **linear `eq_w` scan over every element of self**, so two objects that compare equal but hash differently were treated as the same element — something the table could never do, since `ObjectKey`'s `Hash` writes only `self.hash` and `PartialEq` short-circuits on it. It also meant the toggle mutated self as it went, so a later raise left a partial mutation.

Rather than patch the scan, it now follows `setobject.py:1062-1074 _symmetric_difference_unwrapped`, which does not toggle at all — it builds a **new** storage by walking each side against the other with `contains_with_hash`, then installs it (`:1113-1114`). That fixes the hash-blindness and the partial mutation in one move, and `descr_symmetric_difference` (`:1103-1107`) now shares the same helper.

`{EqBoom()} ^ {EqBoom()}` and `s.symmetric_difference_update(...)` now raise where they previously returned/mutated. Gated in `set_operand_storage.py` (`symdiff_eq_boom`, `ixor_val`).

— *commented by Claude*

File: pyre/pyre-object/src/setobject.rs [hunk 12]
```diff
@@ -174,6 +234,47 @@ pub unsafe fn w_set_discard(obj: PyObjectRef, item: PyObjectRef) -> bool {
...
+/// `obj` must point to a valid `W_SetObject`.
+pub unsafe fn w_set_replace_items(obj: PyObjectRef, items: &[PyObjectRef]) {
+    let s = &mut *(obj as *mut W_SetObject);
+    let entries = &mut *s.items;
+    entries.clear();
+    for &item in items {
+        entries.insert(crate::dictmultiobject::object_key_for(item), ());
+    }
+    s.len = entries.len();
+    set_write_barrier(obj);
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-object/src/setobject.rs]: Agreed on the principle, and the three cited sites are fixed — `w_set_replace_items` is gone, replaced by `w_set_copy_storage_from` (bulk table copy) plus `w_set_key_at`/`w_set_insert_key_checked` for the cases that genuinely compute a new storage. Frozenset conversion and the intersection are answered on their own threads.

One correction to the suggested shape, since it matters. Replacing storage from stored `ObjectKey`s **passed as a slice** is the one thing that is not safe here: `set_object_custom_trace` (pyre-jit/src/eval.rs:564) forwards the `obj` slots **inside** the table, so a `Vec<ObjectKey>` lifted out of a set is invisible to the collector and goes stale across any collection point — including the allocation of the very set it is about to fill. My first draft had exactly that bug (`w_set_from_storage(&storage_copy)`), and it is why the API is now `(dst, src)` over two live sets rather than `(dst, &[ObjectKey])`.

That turns out to be what upstream does anyway: `get_storage_copy:963-966` is `d.copy()` — one bulk copy of the table, no element-by-element refill, no `__hash__`, no `__eq__`. The orthodox shape and the GC-safe shape coincide.

Where a new storage genuinely must be computed (intersection, symmetric difference), the result accumulates into a real allocated set — as `_intersect_unwrapped:1148` accumulates into a GC-managed `get_empty_dict()` — and the source is read one index at a time so it can be re-read after a probe's `eq_w`.

— *commented by Claude*

File: pyre/pyre-interpreter/src/objspace/descroperation.rs [hunk 13]
```diff
@@ -82,6 +82,41 @@ fn bigint_lshift(a: BigInt, shift: usize) -> BigInt {
...
+            drop(probe);
+            ok
+        }
+        Err(_) => false,
+    };
+    if !reservable {
+        return Err(PyError::new(PyErrorKind::MemoryError, ""));
+    }
+    Ok(bigint_lshift(a, shift))
+}
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-interpreter/src/objspace/descroperation.rs:118]: The research is right and I am leaving the code as it is — but the conclusion is a limit on how far this can go, not a defect in the preflight.

malachite exposes no fallible allocation API, so `bigint_lshift` cannot be made to *return* on allocation failure; a genuinely unsatisfiable allocation inside malachite aborts. The `try_reserve_exact` probe does not claim otherwise. What it does is convert the **predictable** case — where the result's limb count is computable up front from `a.bits() + shift` — into a Python `MemoryError`, which is what `1 << 10**18` actually hits and what upstream produces. Without it that case aborted the process.

This is a structural adaptation of the RPython/Rust allocator gap: RPython's allocator is checked, so `rbigint.py:1323 lshift` surfaces MemoryError naturally; Rust's global allocator aborts. The preflight recovers upstream's **observable behaviour** for the reachable inputs, and cannot recover it for genuine OOM under memory pressure — no arrangement of this code could, short of malachite gaining a `try_` surface.

Two notes on the adjacent points already handled: the zero-operand case (`0 << 10**18`) is fixed and returns `0` per `rbigint.py:1323` (`int_other == 0 or selfsign == 0` returns self before sizing anything), and the `usize::try_from` guard exists because `usize` is 32-bit on wasm.

Marking this won't-fix rather than deferring: there is no convergence path to chase until the dependency changes.

— *commented by Claude*

File: pyre/bench/synth/set_update_hash_other.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    m("du_ok", lambda: du({1, 2, 3}, [2]))
+    m("iu_ok", lambda: iu({1, 2, 3}, [2, 3, 4]))
+    m("sdu_ok", lambda: sdu({1, 2}, [2, 3]))
+    m("iu_multi", lambda: iu({1, 2, 3}, [1, 2]))
+    m("du_str", lambda: du(set("abc"), "b"))
+    m("iu_empty_other", lambda: iu({1, 2}, []))
+    m("du_empty_other", lambda: du({1, 2}, []))
+    m("du_int_float", lambda: du({1, 2}, [1.0]))
+    m("iu_two_others", lambda: iu({1, 2}, [1]))
+    m("du_two_others_unhashable", lambda: du({1}, [1]))
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/bench/synth/set_update_hash_other.py:61]: Good catch on both — the cases did not exercise what their names advertised. Rather than make those two lambdas variadic in place, the multi-operand paths are now covered in the new `pyre/bench/synth/set_operand_storage.py`, which is where the operand handling actually lives after this change:

- `intersection_multi` — `a.intersection(b, {3, 2})`
- `difference_multi` — `a.difference({1}, {2})`
- `union_multi` — `a.union({9}, [8])`, mixing a set operand and a non-set one so both branches of `descr_union:365-369` run in one call

and an unhashable/raising element is covered by the `*_eq_boom` lines (`union_eq_boom`, `intersection_eq_boom`, `difference_eq_boom`, `symdiff_eq_boom`, `update_eq_boom`), which now raise rather than returning a container — they were swallowing the `__eq__` before this change.

The whole file is baseline-checked: cpython3.14 and pypy3 agree on every line, so each one is a real gate rather than a snapshot of pyre's own behaviour.

— *commented by Claude*

File: pyre/pyre-object/src/setobject.rs [hunk 15]
```diff
@@ -156,6 +194,28 @@ pub unsafe fn w_set_contains(obj: PyObjectRef, item: PyObjectRef) -> bool {
...
+    item: PyObjectRef,
+) -> Result<bool, crate::dictmultiobject::DictKeyError> {
+    let key = crate::dictmultiobject::object_key_for_checked(item)?;
+    let s = &*(obj as *const W_SetObject);
+    let entries = &*s.items;
+    let found = entries.contains_key(&key);
+    if crate::dictmultiobject::take_dict_key_error() {
+        return Err(crate::dictmultiobject::DictKeyError);
+    }
+    Ok(found)
```
[the remark this answers is not in the document]
Comment (youknowone) [reply pyre/pyre-object/src/setobject.rs:357]: Confirmed and fixed — this one was real, and my own comment on `object_key_hashed` already stated the rule these two were breaking:

> a user `__hash__` is a collection point, and the key caches `obj`, so a key built around the hash call captures the pre-move pointer

`object_key_for_checked` (dictmultiobject.rs:239-247) does exactly that — `try_hash_w(obj)` runs user code, then `ObjectKey { hash, obj }` closes over the **pre-move** `obj`. `w_set_contains_checked` / `w_set_discard_checked` were the two sites that still went through it; the add path had already been converted to hash-then-reload, so this was the leftover half of that conversion.

Both now go through one gate that hashes with the element rooted, reloads it, and probes with the digest (`type_methods.rs set_lookup_checked`), differing only in which primitive they hand the key to — so the two callers stay one line each.

One correction to the advice, since it changes what the fix needs to do: **the set does not need pinning**. `alloc_set_with_type` (setobject.rs:85) allocates through `try_gc_alloc_stable_raw`, whose contract is a pointer "stable across minor and major" (gc_hook.rs:65) — sets are old-gen and keep their addresses. Only the *element* is young and moves. Pinning the set would be harmless but would state something untrue about the object model, so the gate roots the element only and the comment says why.

Being straight about what this does not fix: the `eq_w` inside the bucket probe is itself a collection point, and the key handed to the probe is a stack value the shadow stack does not rewrite — so a probe that spans a relocation can still compare a stale pointer. That is uniform across set and dict (`w_dict_store_object_strategy_checked` has it identically on `dict.__setitem__`), predates this branch, and needs a rooted-key abstraction rather than a per-call-site gate. The change here removes the first-order bug (hash moves the element, key captures the old pointer) and does not pretend to remove the second.

— *commented by Claude*
